consensus/cryptarchia-engine/src/lib.rs

```diff
@@ -252,17 +287,100 @@ where
     pub fn fork_choice(&self) -> Branch<Id> {
         let k = self.config.security_param.get().into();
         let s = self.config.s();
-        Self::maxvalid_bg(self.local_chain.clone(), &self.branches, k, s)
+        Self::maxvalid_bg(self.local_chain, &self.branches, k, s)
     }
 
     pub const fn tip(&self) -> Id {
         self.local_chain.id
     }
 
-    // prune all states deeper than 'depth' with regard to the current
-    // local chain except for states belonging to the local chain
-    pub fn prune_forks(&mut self, _depth: u64) {
-        todo!()
+    pub const fn tip_branch(&self) -> &Branch<Id> {
+        &self.local_chain
+    }
+
+    /// Prune all blocks that are included in forks strictly older than 'depth'
+    /// blocks from the current local chain.
+    ///
+    /// For example, if the tip of the canonical chain is at height 10 (i.e., 11
+    /// blocks long), calling `self.prune_forks(10)` will remove any forks
+    /// stemming from the genesis block, with height `0`, which is the 11th
+    /// block in the past.
+    ///
+    /// This function does not apply any particular logic when evaluating forks
+    /// other than the height at which they diverged from the local
+    /// canonical chain.
+    ///
+    /// It returns the block IDs that were part of the pruned forks.
+    pub fn prune_forks(&mut self, depth: u64) -> impl Iterator<Item = Id> {
+        let prunable_forks = self.prunable_forks(depth);
+        let mut removed_blocks = vec![];
+        for prunable_fork_info in prunable_forks {
+            removed_blocks.extend(self.prune_fork(&prunable_fork_info));
+        }
+        removed_blocks.into_iter()
+    }
+
+    /// Get an iterator over the forks that can be pruned given the provided
+    /// depth.
+    ///
+    /// This means that all forks that diverged from the canonical chain before
+    /// the provided `depth` height are returned.
+    pub fn prunable_forks(&self, depth: u64) -> impl Iterator<Item = ForkDivergenceInfo<Id>> {
+        let local_chain = self.local_chain;
+        let Some(target_height) = local_chain.length.checked_sub(depth) else {
+            tracing::info!(
+                target: LOG_TARGET,
+                "No prunable fork, the canonical chain is not longer than the provided depth. Canonical chain length: {}, provided depth: {}", local_chain.length, depth
+            );
+            return vec![].into_iter();
+        };
+        self.non_canonical_forks()
+            .filter_map(|fork| {
+                // We calculate LCA once and store it in `ForkInfo` so it can be consumed
+                // elsewhere without the need to re-calculate it.
+                let lca = self.branches.lca(&local_chain, &fork);
+                (lca.length <= target_height).then_some(ForkDivergenceInfo { tip: fork, lca })
+            })
+            .collect::<Vec<_>>()
+            .into_iter()
```

Comment:
You do not need to collect. Just let the filter_map. In the other branch simply use [`std::iter::empty`](https://doc.rust-lang.org/std/iter/fn.empty.html)

Reply:
They are still considered two different types, and we can't return two different types as `impl Iterator<Item = ForkDivergenceInfo<Id>>`. This would be solved by `Box<dyn Iterator...>` return type, but I did not deem it necessary here.

```
mismatched types
expected struct `std::iter::Empty<ForkDivergenceInfo<Id>>`
   found struct `std::iter::FilterMap<impl std::iter::Iterator<Item = Branch<Id>> + '_
```

This is the reason for most of the `collect()`, to be honest.